File: CoMa Judge od. Sprechstunde/CoMa_Proggen_HA2.py
class MaxHeap:
    def __init__(self, keys):
        # Erstelle den Heap aus der übergebenen Liste keys.
        self.keys = keys
        n = len(keys)
        # Wende maxHeapify auf jeden Knoten an, der mindestens ein Kind hat.
        for i in range(n//2, -1, -1):
            self.maxHeapify(i, n)
            
    def maxHeapify(self, i, n):
        # Stellt die Max-Heap-Eigenschaft von Knoten i wieder her.
        left = 2*i + 1
        right = 2*i + 2
        largest = i
        # Überprüfe, ob das linke Kind größer als der aktuelle Knoten ist.
        if left < n and self.keys[left] > self.keys[largest]:
            largest = left
        # Überprüfe, ob das rechte Kind größer als der aktuelle Knoten oder das linke Kind ist.
        if right < n and self.keys[right] > self.keys[largest]:
            largest = right
        # Wenn largest != i, tausche die Positionen von i und largest und rufe maxHeapify auf largest auf.
        if largest != i:
            self.keys[i], self.keys[largest] = self.keys[largest], self.keys[i]
            self.maxHeapify(largest, n)
    
    def maximum(self):
        # Das größte Element ist das erste Element in der Liste, da es sich um einen Max-Heap handelt.
        return self.keys[0]
    
    def extractMax(self):
        # Falls der Heap leer ist, gibt es None zurück.
        if len(self.keys) < 1:
            return None
        # Das größte Element ist das erste Element in der Liste.
        max_key = self.keys[0]
        # Ersetze das erste Element mit dem letzten Element in der Liste und entferne das letzte Element.
        self.keys[0] = self.keys[-1]
        self.keys.pop()
        # Stelle die Max-Heap-Eigenschaft wieder her, indem es maxHeapify auf Knoten 0 anwendet.
        self.maxHeapify(0, len(self.keys))
        # Gib das entfernte maximale Element zurück.
        return max_key
    
    def increaseKey(self, i, k):
        # Wenn k kleiner als der aktuelle Schlüssel an Position i ist, kann der Schlüssel nicht erhöht werden.
        if k < self.keys[i]:
            return 'k too small'
        # Andernfalls ersetze den Schlüssel an Position i mit k.
        self.keys[i] = k
        # Führe einen Tausch mit dem Elternteil aus, bis der Elternteil größer als der aktuelle Knoten ist oder der Knoten an der Wurzel angekommen ist.
        while i > 0 and self.keys[(i-1)//2] < self.keys[i]:
            self.keys[(i-1)//2], self.keys[i] = self.keys[i], self.keys[(i-1)//2]
            i = (i-1)//2
    
    def insert(self, k):
        # Füge das Element k am Ende der Liste hinzu.
        self.keys.append(k)
        i = len(self.keys) - 1
        # Führe einen Tausch mit dem Elternteil aus, bis der Elternteil größer als der aktuelle Knoten ist oder der Knoten an der Wur

File: Abgegeben/CoMa_Proggen_HA2.jl
# Definition des Typs MaxHeap
struct MaxHeap
    keys::Vector{Int}
end

# Implementierung der Methoden

# Konstruktor
function MaxHeap(keys::Vector{Int})
    h = new()
    h.keys = keys
    n = length(keys)
    # Aufbau des Max-Heaps
    for i = div(n,2):-1:1
        maxHeapify(h,i)
    end
    return h
end

# maxHeapify - stellt die Max-Heap-Eigenschaft des Teilbaums mit Wurzelknoten i wieder her
function maxHeapify(h::MaxHeap, i::Int)
    n = length(h.keys)
    l = 2i # linker Kindknoten
    r = 2i+1 # rechter Kindknoten
    largest = i
    if l <= n && h.keys[l] > h.keys[i]
        largest = l
    end
    if r <= n && h.keys[r] > h.keys[largest]
        largest = r
    end
    if largest != i
        h.keys[i], h.keys[largest] = h.keys[largest], h.keys[i]
        maxHeapify(h,largest)
    end
end

# maximum - gibt das maximale Element des Max-Heaps zurück
function maximum(h::MaxHeap)
    return h.keys[1]
end

# extractMax - gibt das maximale Element des Max-Heaps zurück und stellt die Max-Heap-Eigenschaft wieder her
function extractMax(h::MaxHeap)
    n = length(h.keys)
    if n < 1
        error("Heap underflow")
    end
    max = h.keys[1]
    h.keys[1] = h.keys[n]
    deleteat!(h.keys,n)
    maxHeapify(h,1)
    return max
end

# increaseKey - erhöht den Eintrag von keys[i] auf k, falls k größer ist als keys[i], und stellt anschließend die Max-Heap-Eigenschaft wieder her
function increaseKey(h::MaxHeap, i::Int, k::Int)
    if k < h.keys[i]
        error("New key is smaller than current key")
    end
    h.keys[i] = k
    while i > 1 && h.keys[div(i,2)] < h.keys[i]
        h.keys[i], h.keys[div(i,2)] = h.keys[div(i,2)], h.keys[i]
        i = div(i,2)
    end
end

# insert - fügt ein Element mit dem Schlüssel k in keys ein und stellt anschließend die Max-Heap-Eigenschaft wieder her
function insert(h::MaxHeap, k::Int)
    push!(h.keys,k)
    n = length(h.keys)
    i = n
    while i > 1 && h.keys[div(i,2)] < h.keys[i]
        h.keys[i], h.keys[div(i,2)] = h.keys[div(i,2)], h.keys[i]
        i = div(i,2)
    end
end

# heapSort - führt Heapsort auf dem Max-Heap durch, so dass keys aufsteigend sortiert wird
function heapSort(h::MaxHeap)
    n = length(h.keys)
    for i = n:-1:2
        h.keys[1], h.keys[i] = h.keys[i], h.keys[1]
        n -= 1
        maxHeapify(h,1)
    end
end
